src/lib.rs

```diff
@@ -5,10 +5,32 @@
 
 extern crate vm_memory;
 
-use vm_memory::GuestAddress;
+use std::cmp::{Ord, Ordering, PartialOrd};
+use vm_memory::{Address, GuestAddress, GuestUsize};
 
+pub mod device_manager;
 pub mod resources;
 
+/// IO Size.
+#[derive(Debug, Copy, Clone)]
+pub enum IoSize {
+    /// Port I/O size.
+    Pio(u16),
+
+    /// Memory mapped I/O size.
+    Mmio(GuestUsize),
+}
+
+impl IoSize {
+    /// Get the raw value as u64 to make operation simple.
+    pub fn raw_value(&self) -> u64 {
+        match *self {
+            IoSize::Pio(p) => u64::from(p),
+            IoSize::Mmio(m) => m,
```

Comment:
This assumes that GuestUsize is u64:)

Reply:
> This assumes that GuestUsize is u64:)

It seems that vm-memory definition shows GuestUsize is u64.
``` rust
impl_address_ops!(GuestAddress, u64);

macro_rules! impl_address_ops {
    ($T:ident, $V:ty) => {
        impl AddressValue for $T {
            type V = $V;
        }

```
```
pub type GuestUsize = <GuestAddress as AddressValue>::V;
```

Reply:
That's an vm-memory internal detail, it would be better to rely on that.

Reply:
Actually I'm hesitating whether we should remove dependency on the vm-memory crate by using u64 instead of GuestAddress/GuestUsize.
What do you think?

Reply:
> Actually I'm hesitating whether we should remove dependency on the vm-memory crate by using u64 instead of GuestAddress/GuestUsize.
> What do you think?

When we considering using vm-memory::GuestAddress/GuestUsize or not, it means two things: whether IoManager knows the mmio address type, and whether device IO interface (`DeviceIo::read/write`) and device inside knows the mmio address type (for resource requirement and assignment). It would be better to keep those the same.

Reply:
> > Actually I'm hesitating whether we should remove dependency on the vm-memory crate by using u64 instead of GuestAddress/GuestUsize.
> > What do you think?
> 
> When we considering using vm-memory::GuestAddress/GuestUsize or not, it means two things: whether IoManager knows the mmio address type, and whether device IO interface (`DeviceIo::read/write`) and device inside knows the mmio address type (for resource requirement and assignment). It would be better to keep those consistent.

So we need to decide the argument to DeviceIo::{read,write} is an address or an offset.
I suggest to use offset, because that will simplify the driver implementations.

Reply:
I also prefer the argument to deviceio to be an offset. In which case I would also propose to change the signature of read/write to use `offset: u64` instead of `addr: IoAddress`.

Reply:
> I also prefer the argument to deviceio to be an offset. In which case I would also propose to change the signature of read/write to use `offset: u64` instead of `addr: IoAddress`.

We still need a way to encode the access type: PIO vs MMIO. So we either add an extra `access_type` parameter or use the IoAddress enum. I suggest to use IoAddress enum to encode the access type. The reason is that PCI device need to support both PIO and MMIO.

Reply:
> We still need a way to encode the access type: PIO vs MMIO. So we either add an extra `access_type` parameter or use the IoAddress enum. 

Agree. In my very previous PR https://github.com/rust-vmm/vm-device/pull/3/commits/538836161c3dddb37f268ca41e1fa565ebfa7492,  there proposed such a type differing pio and mmio. 
> I suggest to use IoAddress enum to encode the access type. The reason is that PCI device need to support both PIO and MMIO.

Encoding it into address type is a good way to simply interface. However, once we choose current `IoAddress` enum for `DeviceIo::read/write`, it means `vm_device::IoManager` cares `GuestAddress/GuestUsize` instead of direct type u64, and transports such semantics to virtual device. So virtual device has semantics of `GuestAddress/GuestUsize` instead of direct u64 for IO windows, which means `ResourceConstraint` and `Resource` also need that semantics too.  That is what I mean by keeping consistency.

One simple way is, only use u64 instead of `GuestAddress`, and encode in enum as `offset` for IoManager::read/write.
```rust
/// IO Addresses.
#[derive(Debug, Copy, Clone)]
pub enum IoAddress {
    /// Port I/O address.
    Pio(u16),

    /// Memory mapped I/O address.
    Mmio(u64),
}
```

Reply:
I like this design, make vm-device independent of vm-memory.

Reply:
> I like this design, make vm-device independent of vm-memory.

I like it as well. That combined with replacing the current `addr` argument with `base + offset` (see https://github.com/rust-vmm/vm-device/pull/13/files#r364223789) could make it into the same PR, to be merged before that one.